# Statistics and Probability

## 1. Descriptive Statistics

In [44]:
import pandas as pd
import numpy as np
import matplotlib as mlt
Iris=pd.read_csv('./Resources/Iris.csv')
Iris.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
0   1            5.1           3.5            1.4           0.2
1   2            4.9           3.0            1.4           0.2
2   3            4.7           3.2            1.3           0.2
3   4            4.6           3.1            1.5           0.2
4   5            5.0           3.6            1.4           0.2

a. Compute the mean, median, and standard deviation for all numerical features.  

* Mean

In [45]:
SepalLengthCm_mean = Iris.SepalLengthCm.mean()
SepalWidthCm_mean = Iris.SepalLengthCm.mean()
PetalLengthCm_mean = Iris.PetalLengthCm.mean()
PetalWidthCm_mean =Iris.PetalWidthCm.mean()
print(f'SepalLengthCm_mean={SepalLengthCm_mean}\nSepalWidthCm_mean={SepalWidthCm_mean}\nPetalLengthCm_mean={PetalLengthCm_mean}\nPetalWidthCm_mean={PetalWidthCm_mean}')

SepalLengthCm_mean=5.843333333333334
SepalWidthCm_mean=5.843333333333334
PetalLengthCm_mean=3.758666666666666
PetalWidthCm_mean=1.1986666666666668


* Median

In [46]:
SepalLengthCm_median = Iris.SepalLengthCm.median()
SepalWidthCm_median = Iris.SepalLengthCm.median()
PetalLengthCm_median = Iris.PetalLengthCm.median()
PetalWidthCm_median =Iris.PetalWidthCm.median()
print(f'SepalLengthCm_median={SepalLengthCm_median}\nSepalWidthCm_median={SepalWidthCm_median}\nPetalLengthCm_median={PetalLengthCm_median}\nPetalWidthCm_median={PetalWidthCm_median}')

SepalLengthCm_median=5.8
SepalWidthCm_median=5.8
PetalLengthCm_median=4.35
PetalWidthCm_median=1.3


* Standard Deviation

In [47]:
SepalLengthCm_sd = Iris.SepalLengthCm.std()
SepalWidthCm_sd = Iris.SepalLengthCm.std()
PetalLengthCm_sd = Iris.PetalLengthCm.std()
PetalWidthCm_sd =Iris.PetalWidthCm.std()
print(f'SepalLengthCm_sd={SepalLengthCm_sd}\nSepalWidthCm_sd={SepalWidthCm_sd}\nPetalLengthCm_sd={PetalLengthCm_sd}\nPetalWidthCm_sd={PetalWidthCm_sd}')

SepalLengthCm_sd=0.828066127977863
SepalWidthCm_sd=0.828066127977863
PetalLengthCm_sd=1.7644204199522626
PetalWidthCm_sd=0.7631607417008411


 b. Create histograms and boxplots for feature distributions.

 c. Calculate the correlation matrix for the features.

In [48]:
Matrix=Iris.corr()
Matrix

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  \
Id             1.000000       0.716676     -0.397729       0.882747   
SepalLengthCm  0.716676       1.000000     -0.109369       0.871754   
SepalWidthCm  -0.397729      -0.109369      1.000000      -0.420516   
PetalLengthCm  0.882747       0.871754     -0.420516       1.000000   
PetalWidthCm   0.899759       0.817954     -0.356544       0.962757   

               PetalWidthCm  
Id                 0.899759  
SepalLengthCm      0.817954  
SepalWidthCm      -0.356544  
PetalLengthCm      0.962757  
PetalWidthCm       1.000000

 d. Interpret which features are most correlated with each other.

The most correlated features are PetalLength and PetalWidth with correlation coefficient of 0.9627.

## 2. Conditional Probability

Given the following table of events, compute P(A∣B) and P(B∣A).<br>
<img src='./Resources/Question-2-table.png'>


In [49]:
A_and_B = 30  
A_and_not_B = 20  
not_A_and_B = 50  
not_A_and_not_B = 100  
total_A = A_and_B + A_and_not_B
total_B = A_and_B + not_A_and_B
P_A_given_B = A_and_B / total_B
P_B_given_A = A_and_B / total_A
print(f"P(A | B) = {P_A_given_B}")
print(f"P(B | A) = {P_B_given_A}")

P(A | B) = 0.375
P(B | A) = 0.6


## 3. Bayes' Theorem

A model classifies spam emails with the following probabilities:<br>
    P(Spam)=0.2<br>
    P(Not Spam)=0.8<br>
    P(Keyword | Spam)=0.7<Br>
    P(Keyword | Not Spam)=0.1<br>

Compute the probability that an email is spam given that it contains the keyword.

In [50]:
P_spam = 0.2
P_not_spam = 0.8
P_keyword_given_spam = 0.7
P_keyword_given_not_spam = 0.1
P_keyword = (P_keyword_given_spam * P_spam) + (P_keyword_given_not_spam * P_not_spam)
P_spam_given_keyword = (P_keyword_given_spam * P_spam) / P_keyword
print(f"P(Spam | Keyword) = {P_spam_given_keyword:.3f}")

P(Spam | Keyword) = 0.636


## 4. Outlier Detection, Removal, and Feature Scaling

Choose a dataset on your own and perform the following operation
* Compute Q1 (25th percentile) and Q3 (75th percentile) for each feature.
* Calculate the IQR (IQR=Q3−Q1).
* Define outliers as values outside the range [Q1−1.5⋅IQR,Q3+1.5⋅IQR].
* Create a function to detect and remove outliers for all numerical features.
* Plot boxplots of features before and after outlier removal.  
* Plot histograms of a few features before and after applying normalization and standardization.